eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/ImmutableSetFromBagAdapter.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020 Goldman Sachs and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v. 1.0 which accompany this distribution.
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ */
+
+package org.eclipse.collections.impl.bag.immutable;
+
+import java.util.Iterator;
+import java.util.Set;
+
+import org.eclipse.collections.api.bag.Bag;
+import org.eclipse.collections.api.block.procedure.Procedure;
+import org.eclipse.collections.api.set.ImmutableSet;
+import org.eclipse.collections.impl.set.immutable.AbstractImmutableSet;
+
+/**
+ * This class provides a {@link ImmutableSet} wrapper around any Bag interface instance.
+ */
+public final class ImmutableSetFromBagAdapter<T> extends AbstractImmutableSet<T>
+{
+    private final Bag<T> delegate;
+
+    public ImmutableSetFromBagAdapter(Bag<T> delegate)
+    {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public int size()
+    {
+        return this.delegate.sizeDistinct();
+    }
+
+    @Override
+    public T getFirst()
+    {
+        return this.delegate.getFirst();
+    }
+
+    @Override
+    public T getLast()
+    {
+        return this.delegate.getLast();
+    }
+
+    @Override
+    public void each(Procedure<? super T> procedure)
+    {
+        this.delegate.distinctView().each(procedure);
+    }
+
+    @Override
+    public Iterator<T> iterator()
+    {
+        return this.delegate.distinctView().iterator();
```

Comment:
Is there a test to make sure that this iterator's remove() method throws?

Reply:
@motlin good point, I added tests for the new classes (by extending the set test classes), it enabled me to find a few bugs too :)